### ISSUE TITLE
Seeking clarification on state transitions

### ISSUE THREAD

**Issue by misberner** [helm/helm#9843]
Hi there,

in the context of the [helm-operator](https://github.com/operator-framework/helm-operator), we recently had [a discussion](https://github.com/operator-framework/helm-operator) ([more](https://github.com/operator-framework/helm-operator), [more](https://github.com/stackrox/helm-operator/pull/13#issuecomment-865112083), [more](https://github.com/stackrox/helm-operator/pull/13#issuecomment-865146142)) around the precise semantics of a `-pending` status.

The question boils down to the following: we are looking for a confirmation that
- **if** a release is in a `-pending` state,
- **and** I know that there currently is no active/live Helm client that performs a Helm action,
- **then I can assume** that a rollback is the only course of action that will allow me to keep using the release (w/o deleting and recreating it)

is a valid reasoning for at least all versions of Helm 3 (including future versions). I am aware of the practical implications of doing a rollback, especially in the presence of hook jobs, but these seem orthogonal. Basically, what I want to know is if the authors of this project can imagine that in the future, a `-pending` state will be used in such a way that it is the final state of a release after a **successful** Helm action, rather than a failed/interrupted/... one only, as is currently the case. (Not necessarily realistic) examples of that could be: 1) explicit transaction commands (`helm begin-upgrade ; ... helm commit-upgrade`), 2) explicit release locking (`helm lock ; helm unlock`).

Really appreciate your input here.

**Comment by bacongobbler** [comment]
The short answer: there should not be a case where a release is in a `pending` state and there are no active/live Helm clients.

If you enter this state, one cannot assume that a rollback is the only course of action. You have entered what most state machines call an "Unknown" state, where any future operations are unknown. A `rollback` cannot guarantee that it will fix the release. In many cases you could make the situation worse by taking down production services, or irrevocably removing production data due to a release hook running a database migration.

So no I would not assume that a `helm rollback` is a valid course of action. If you enter this state, you have entered an erroneous case and the user should intervene to determine next steps.

Hope this helps answer your question.

**Comment by misberner** [comment]
@bacongobbler thank you for your response. I am fully aware of the risks that doing a rollback could have in the presence of hook jobs. I did not intend "Only course of action" to mean "the right thing to do", but instead "on the Helm mechanics side, other than deleting the release I will not be able to successfully make any modifications on the release via upgrade unless I first execute a rollback". This seems to be strictly enforced in `helm upgrade`. The only other option could be to manually edit the release secret to change the `-pending` state into a non-pending one, but that seems even more problematic than a rollback.

**Comment by hickeyma** [comment]
@misberner You could handle it as follows:

- If the objects are all deployed in the cluster for the release version, then you could update the state of the release version to `deployed` (release version metadata stored as secret in the cluster)
- If there has been no update of the objects deployed in the cluster from the previous release version, then you could delete the release version secret in the cluster and update the state of the previous release version  to `deployed`
- If some not all of the changes have been deployed to the cluster .. not sure the best approach to this bar delete the release and maybe cleanup the cluster where necessary

**Comment by bacongobbler** [comment]
Yeah... I think @hickeyma is onto something. That being said, you're now having to understand the differences between "no objects have been updated" and "all objects have been updated, but current state is `pending`". I'm not sure we can provide a workflow that agrees to the following criteria:

> is a valid reasoning for at least all versions of Helm 3 (including future versions).

Because how Helm upgrades resources and how Helm determines who and what has been updated may change from version to version. Helm 2 and Helm 3 saw one such major architectural change that now includes the live state via a 3-way strategic merge patch, and we've discussed in the past using Kubernetes' server-side apply API in place of Helm's client-side upgrade logic. We also saw the release version metadata change pretty significantly between Helm 2 and Helm 3.

In summary, @hickeyma's reasoning could be reasonable for Helm 3. We can make certain logical guarantees for ONE major version. But we cannot make guarantees that Helm's state logic will be forwards-and-backwards-compatible for all major Helm versions, in perpetuity. :)

**Comment by hickeyma** [comment]
I agree with what @bacongobbler said above. I don't think that it is something that should be implemented in core.

It could be something done as a stand-alone program or Helm plugin. It might be worth looking at the following for examples:

- [2to3 plugin](https://github.com/helm/helm-2to3) which performs essentially ETL of the Helm v2 release metadata in the cluster to Helm v3 release metadata.
- [mapkubeapis plugin](https://github.com/hickeyma/helm-mapkubeapis) which updates the manifest data stored in the Helm release metadata in the cluster.

Both these plugins are essential tools to Helm but should not be part of Helm core because they don't add to the core capability of the product and would bloat the code.

**Comment by misberner** [comment]
So when I said "Helm 3 (including future versions)", I did not mean 4+, but 3.X for an arbitrary X.

Practically, I do agree with the first two points. But is the format of a Helm release actually part of documented API with any stability and not only backward but forward-compatibility guarantees? I.e., will the data always be stored in the `release` entry in the secret, will the JSON path always be `info.status`, will `deployed` always remain a valid state? Note, I'm not looking for guarantees that this will be such for all eternity. But at least in my understanding, "simple" backwards-compatibility is not enough to guarantee these.

Second, and because of the former, I obviously have concerns when it comes to manipulating such data on a stored secret. These could mostly be mitigated by using the `Secrets` driver or the `Storage` API to get and then update the release, but it still seems odd w/o a CLI for this. Also, even if there was a CLI, it would be something like `helm set-release-state --force`. This does not seem better than a rollback, in terms of safety.

Finally, this discussion makes me wonder if there wasn't maybe a place for a `helm <upgrade|rollback> --fix` or `helm <upgrade|rollback> --force-continue`. This would basically work on a locked release, and attempt to continue the most recent operation, using the rendered manifests already stored in the secret:
- conservatively (see below) assume that pre-hooks haven't been executed, and re-execute them, if there are any.
- make all kubernetes resources under management equal `Manifest` stored in the release (nothing changes for any resources where this has already been applied)
- run all post-hooks

It would certainly be possible to make sure that the pre-hook/manifest update/post-hook phases each run in strict sequence, by introducing intermediate pending states (`upgrade-pending; upgrade-pending-pre-hooks-finished; upgrade-pending-manifests-applied; deployed`). However, each of the phases should be robust to restarts and concurrent executions, because Kubernetes itself doesn't provide stronger guarantees: I cannot tell a `Job` to only attempt to schedule a single pod - if it gets evicted, the Job controller will create a new one on a different node, I can only indirectly (and imperfectly) prevent that through a tight deadline. Similarly, a kubelet failure may effectively remove a node from the cluster and cause a pod to get rescheduled, even though the container keeps running on the original node, possibly accessing the same shared NFS storage that is also used by the newly scheduled replacement pod. The manifest application phase, finally, is all about desired state anyway, application of which should be idempotent.

I can see the devastating effects of running a rollback (and its hooks) concurrently with an upgrade (and its hooks). But I think if an app is not robust against concurrent upgrades with the same target values/manifests, it is not robust against real-life cluster malfunction scenarios in general.

**Comment by bacongobbler** [comment]
> So when I said "Helm 3 (including future versions)", I did not mean 4+, but 3.X for an arbitrary X.

Ah. In that case, yes, this behavior can be expected to remain consistent from 3.0.0 to the end of its service life.
